Player lifecycle docs update (#820)
spatialos/gdk-for-unity: docs/content/modules/player-lifecycle-feature-module.md

The Player lifecycle module is a feature module providing you with a simple player spawning and player lifecycle management implementation.

> ## How to enable this module

Reviewer: intentional?

Reviewer 2: I think so - it looks like it was off by one originally.

To enable this module, you need to add the necessary systems to your workers in the `HandleWorkerConnectionEstablished()` of your [`WorkerConnector`]({{urlRoot}}/content/gameobject/creating-workers-with-workerconnector).
After your worker has been created, you need to use the following code snippet:

1. On a [client-worker]({{urlRoot}}/content/glossary#client-worker): `PlayerLifecycleHelper.AddClientSystems(Worker.World)`
  * By default, the module automatically sends a player creation request as soon as the client worker connects to SpatialOS. To manually initiate player creation or provide arbitrary data in a player creation request, add an extra `false` argument when adding client systems: `PlayerLifecycleHelper.AddClientSystems(Worker.World, false);`
1. On a [server-worker]({{urlRoot}}/content/glossary#server-worker): `PlayerLifecycleHelper.AddServerSystems(Worker.World)`

The Player lifecycle module provides a `PlayerLifecycleConfig` configuration class, containing a field called `AutoRequestPlayerCreation`. By default this is set to true, in which case the client-worker initiates the player entity creation as soon as it has connected to SpatialOS.

> Calling `PlayerLifecycleHelper.AddClientSystems(Worker.World, false)` sets this value to false, in which case you must manually initiate player creation.

Reviewer 2: Why is this configuration info here and is it strictly necessary? 

As a user I just need to know that if I want to manually send the request, to set the optional argument to false.

## How to spawn a player entity

To spawn a [SpatialOS entity]({{urlRoot}}/content/glossary#spatialos-entity) representing
a player, you need to:

1. Define the entity template for your player entity
1. Specify an entity template for player creation
1. (Optional) Manually request player creation

> >**NOTE: To pass in arbitrary serialized data with the player creation request, `AutoRequestPlayerCreation` must be set to false.**

Reviewer 2: Not sure if our type is all caps NOTE

The Player lifecycle module is a feature module providing you with a simple player spawning and player lifecycle management implementation.

## How to enable this module

To enable this module, you need to add the necessary systems to your workers in the `HandleWorkerConnectionEstablished()` of your [`WorkerConnector`]({{urlRoot}}/content/gameobject/creating-workers-with-workerconnector).
After your worker has been created, you need to use the following code snippet:

1. On a [client-worker]({{urlRoot}}/content/glossary#client-worker): `PlayerLifecycleHelper.AddClientSystems(Worker.World)`

>   * By default, the module automatically sends a player creation request as soon as the client worker connects to SpatialOS. To manually initiate player creation or provide arbitrary data in a player creation request, add an extra `false` argument when adding client systems: `PlayerLifecycleHelper.AddClientSystems(Worker.World, false);`

Reviewer 3: ```suggestion
  * By default, the module automatically sends a player creation request as soon as the client-worker instance connects to SpatialOS. To manually initiate player creation or provide arbitrary data in a player creation request, add an extra `false` argument when adding client systems: `PlayerLifecycleHelper.AddClientSystems(Worker.World, false);`
```
Pedantic. Being specific about using "instance" now. :-)

1. On a [server-worker]({{urlRoot}}/content/glossary#server-worker): `PlayerLifecycleHelper.AddServerSystems(Worker.World)`

## How to spawn a player entity

To spawn a [SpatialOS entity]({{urlRoot}}/content/glossary#spatialos-entity) representing
a player, you need to:

1. Define the entity template for your player entity
1. Specify an entity template for player creation
1. (Optional) Manually request player creation

> >**NOTE: To pass in arbitrary serialized data with the player creation request, `AutoRequestPlayerCreation` must be set to false.**

Reviewer 3: ```suggestion
>**NOTE:** To pass in arbitrary serialized data with the player creation request, set `AutoRequestPlayerCreation` to false.
```
Assume that it's something our user has to do?  If not and it's just a check, "make sure set `AutoRequestPlayerCreation` is set to false."

If `AutoRequestPlayerCreation` is set to false you must manually initiate player creation, as described in Step 3.

##### 1. Define the player entity template

The module takes care of spawning the player entity as soon as a client-worker connects. To enable this behaviour, the server-worker responsible for handling requests to spawn player entities needs to know which [entity template]({{urlRoot}}/content/entity-templates) to use when sending the entity creation request to the [SpatialOS Runtime]({{urlRoot}}/content/glossary#spatialos-runtime).

Create a method that returns an `EntityTemplate` object and takes the following parameters to define your player [entity template]({{urlRoot}}/content/entity-templates):

* `string workerId`: The ID of the worker that wants to spawn this player entity.

> *  `byte[] playerCreationArguments`: a serialized byte array of arguments provided by the worker sending the player creation request.

Reviewer 3: ```suggestion
*  `byte[] playerCreationArguments`: a serialized byte array of arguments provided by the worker isntance sending the player creation request.
```

> When defining the entity template, you need to use the `AddPlayerLifecycleComponents` method. This methods adds the SpatialOS components to the player entity template necessary to manage the player lifecycle for this entity.

Reviewer 3: ```suggestion
When defining the entity template, you need to use the `AddPlayerLifecycleComponents` method. This method adds lifecycle SpatialOS components to the player entity template; these components are necessary to manage the player lifecycle for this entity.
```
Just tried to make the last sentence flow without stumbling - original took a few reads to follow.

}
```

##### 2. Specify an entity template for player creation

> The `PlayerLifecycleConfig` configuration class needs to be configured to use the player entity template. This can be done by setting the `PlayerLifecycleConfig.CreatePlayerEntityTemplate` field to `PlayerTemplate.CreatePlayerEntityTemplate`.

Reviewer 3: ```suggestion
You need to configure the `PlayerLifecycleConfig` configuration class to use the player entity template. Do this by setting the `PlayerLifecycleConfig.CreatePlayerEntityTemplate` field to `PlayerTemplate.CreatePlayerEntityTemplate`.
```
Made it active voice. (It was passive voice in the original.)

        var serverAttribute = "UnityGameLogic";

        var deserializedArguments = DeserializeArguments<SampleArgumentsObject>(playerCreationArguments);

        var entityTemplate = new EntityTemplate();
        entityTemplate.AddPosition(new Position.Snapshot { Coords = deserializedArguments.SpawnPosition }, serverAttribute);
        entityTemplate.AddComponent(new PlayerName.Snapshot { PlayerName = deserializedArguments.PlayerName }, serverAttribute);
        // add all components that you want the player entity to have
        AddPlayerLifecycleComponents(entityTemplate, workerId, clientAttribute, serverAttribute);

        return entityTemplate;
    }

    // An example implementation of DeserializeArguments
    private T DeserializeArguments<T>(byte[] serializedArguments)
    {
        using (var memoryStream = new MemoryStream())
        {
            var binaryFormatter = new BinaryFormatter();
            memoryStream.Write(serializedArguments, 0, serializedArguments.Length);
            memoryStream.Seek(0, SeekOrigin.Begin);
            return (T) binaryFormatter.Deserialize(memoryStream);
        }
    }
}
```

## When is a player entity deleted?

> To ensure that player entities of disconnected client-workers get deleted correctly, the server-workers responsible to manage the player lifecycle sends a `PlayerHeartBeat` [command]({{urlRoot}}/content/world-component-commands-requests-responses) to the different player entities to check whether they are still connected. If a player entity fails to send a response three times in a row, the server-worker sends a request to the SpatialOS Runtime to delete this entity.

Reviewer 3: ```suggestion
To ensure that player entities of disconnected client-workers instances get deleted correctly, the server-worker instances responsible for managing the player lifecycle sends a `PlayerHeartBeat` [command]({{urlRoot}}/content/world-component-commands-requests-responses) to the different player entities to check whether they are still connected. If a player entity fails to send a response three times in a row, the server-worker instance sends a request to the SpatialOS Runtime to delete this entity.
```
The sentence was a bit iffy. Also added instance coz that's how we role now.

}
```

##### 3. (Optional) Manually request player creation

If `AutoRequestPlayerCreation` is set to false, you must manually call `RequestPlayerCreation` in the `SendCreatePlayerRequestSystem`:

> ```csharp

Reviewer 2: Nit: Line of whitespace between text and code

var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
playerCreationSystem.RequestPlayerCreation();
```

> ## How to pass arbitrary data into a player creation request

Reviewer 2: Maybe: "How to use arbitrary data in the player creation loop" or something like that. This covers it end to end really

##### 1. Set AutoRequestPlayerCreation to false

> First, ensure that `AutoRequestPlayerCreation` is set to false.

Reviewer 2: Please put whitespace between distinct Markdown elements. Just because Improbadoc doesn't care doesn't mean I don't 😛

> ##### 2. Serialize before calling RequestPlayerCreation

Reviewer 2: Temporal part of this title feels a bit strange. And what are we serializing?

Then, ensure that it is serialized into a byte array before calling `RequestPlayerCreation`:
```csharp
var myArguments = new SampleArgumentsObject { PlayerName = "playerName", SpawnPosition = new Coordinates(42, 0, 69)) };
var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
var serializedArguments = PlayerLifecycleHelper.SerializeArguments(myArguments);
playerCreationSystem.RequestPlayerCreation(serializedArguments);

> ```

Reviewer 2: Nit: Whitespace between code and header
